### PR TITLE
Drop 0.3 and support 0.5

### PR DESCRIPTION
**Pull request by iamed2**
🎉 

Some `Union{}[]` and `Any[]` and `Any[...]` were changed to `[]` and `[...]`. 

v0.3 support was dropped per the METADATA freeze. Code that only runs on versions before the v0.4.0 release was deleted, and unnecessary `Compat` usage was deleted. The old `test/test.jl` file was deleted as its contents are replicated/improved in `test/runtests.jl` (the `test/perf.jl` wasn't removed, as it is not yet brought up to date). 

v0.5 support was fixed and tested.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -271,7 +275,7 @@ end
 
 @test_zip [1,2,3] [:a, :b, :c] ['x', 'y', 'z']
 @test_zip [1,2,3] [:a, :b] ['w', 'x', 'y', 'z']
-@test_zip [1,2,3] @compat(Union{})[] ['w', 'x', 'y', 'z']
```
Comment (yuyichao) [inline test/runtests.jl]: Why are these changed?

Comment (iamed2) [inline test/runtests.jl]: The type of the empty array is not material to the test. It was only changed from `[]` to avoid a deprecation warning from https://github.com/JuliaLang/julia/pull/8578.

Comment (yuyichao) [inline test/runtests.jl]: `[]` is `Union{}[]` on 0.3. And this feels like a corner case that still worth testing.

Comment (iamed2) [inline test/runtests.jl]: Alright, testing both now.

File: test/runtests.jl [hunk 2]
```diff
@@ -122,26 +126,26 @@ end
 
 # Empty arrays
 test_imap(
-  Any[],
```
Comment (yuyichao) [inline test/runtests.jl]: Also these.

Comment (iamed2) [inline test/runtests.jl]: This is really not designed to test that. What adding those would really be testing is `Union{}[] == Any[]` which is not in the scope of Iterators.

Comment (yuyichao) [inline test/runtests.jl]: Well, those test was written in that way to begin with.

It's hard to say if it's in scope of Iterators but it doesn't hurt to not change the tests.

File: test/runtests.jl [hunk 3]
```diff
@@ -126,26 +126,26 @@ end
 
 # Empty arrays
 test_imap(
-  [],
```
Comment (yuyichao) [inline test/runtests.jl]: This and the last one is still wrong.

File: test/runtests.jl [hunk 4]
```diff
@@ -294,7 +299,7 @@ macro test_enumerate(input)
 end
 
 @test_enumerate [:a, :b, :c]
-@test_enumerate @compat(Union{})[]
+@test_enumerate []
```
Comment (yuyichao) [inline test/runtests.jl]: Also, here..... I was trying to not leave a comment at every single place.

Comment (iamed2) [inline test/runtests.jl]: Ugh, sorry. Missed that one in the first batch.

File: test/runtests.jl [hunk 5]
```diff
@@ -299,7 +309,8 @@ macro test_enumerate(input)
 end
 
 @test_enumerate [:a, :b, :c]
-@test_enumerate []
+@test_enumerate Union{}[]
+@test_enumerate Any[]
```
Comment (yuyichao) [inline test/runtests.jl]: Note that on 0.4+ `Any[]` is the same with `[]`. Doesn't really mater which one to use.

Comment (iamed2) [inline test/runtests.jl]: Yep. Just being more explicit as these tests are no longer just meant to test "empty arrays".
